### DISCUSSION TITLE
Adding branches after successfully ran one

### DISCUSSION THREAD

**Discussion by llrs** [ropensci/targets#834] [Help] [answered]
Apologies for this long post, it describes the situation I found myself, some solutions I thought and some problems I found with them. Many thanks for the amazing package.

# Background

I have a target something like this:

```r
list(
    tar_target(A, do_something()), 
    tar_target(B, do_something2(A)),
    tar_target(C, do_something3(B, A)),
    tar_target(D, do_something4(C))
)
```

All these targets are already finished and tracked by targets. 

Now I need to exclude some samples and repeat the process again (Some quality issues on the original data I didn't caught earlier), but I do not want to lose all the current data. I could add an intermediate step and copy and repeat the targets (with a new name) as follows:

```r
list(
    tar_target(A, do_something()), 
    tar_target(B, do_something2(A)),
    tar_target(C, do_something3(B, A)),
    tar_target(D, do_something4(C))
   #  Modify A and repeat all process
    tar_target(A2, filter_input(do_something(), D)), 
    tar_target(B2, do_something2(A2)),
    tar_target(C2, do_something3(B2, A2)),
    tar_target(D2, do_something4(C2))
)
```
However there are close to 230 targets and this could be a mess to do it manually. 

# Solutions?

I've found/came up with two options:

1) Use branching

If I understood correctly this means dynamic branching (very well [explained on the book](https://books.ropensci.org/targets/dynamic.html)). 
However, my own filtering depends on some targets downstream where I would add pattern (C2 for instance).

```r
library("targets")
mislabeled <- iris[ iris$Species == "Setosa", ]
tar_pattern(slice(x, index = which(!iris$Species %in% mislabeled$Species)), x = nrow(iris)) # Dynamic branching but one branch for each row of iris?
# Add a new slice?
tar_pattern(slice(x, index = list(sub = which(!iris$Species %in% mislabeled$Species), all = seq_len(nrow(iris))), x = nrow(iris))) # This one fails
```

On the code above I would replace the which(...) by those indexes c(1L, 57L, ...150L). Which then might be better if I use static branching

Maybe I'm focusing on the wrong problem and I should use a list:

```r
l <- list(all = iris, subset = iris[!iris$Species == "Setosa", ])
tar_pattern(map(l), l = 2)
```
However, If I replace the the original target (A in the examples above) with something like this then targets would need to rerun the branch with the whole dataset (and it took 2 weeks to finish the workflow...) and the one with the subset. 

Perhaps static branching would avoid redoing this target, but my understanding of it is that it should be planned from the start. 

2) I could save elsewhere the current targets, do all the process again and read the original output again (manually). This would be similar to doing everything manually. But that would be my last resource. 

# Wrapping up

Maybe I missed something but there doesn't seem to be anything aside from manually creating the new "branch" that would allow me to keep the current targets build. Is there a branching extension mechanism? 

TLDR: I need to redo a long target, adding a "branch". Current branching schemes do not seem to save time to redo the work when  "a branch" is already finished. Anything I might have missed?

**Comment by wlandau** [comment, marked as answer]
Seems like you want to preserve the historical record of your prior targets. [`gittargets`](https://docs.ropensci.org/gittargets/) is one approach, and another is [cloud storage](https://books.ropensci.org/targets/data.html#cloud-storage) with `_targets/meta/meta` committed to Git and versioning turned on in the bucket. I think either of those would be more convenient for you than branching.

I'm actually going to give a short talk on this side of `targets` at useR!2022: https://wlandau.github.io/user-conf-2022/#1

**Reply by llrs** [reply]
Many thanks, I read about gittargets before but I thought it was only for (input) data, now I realize it solves exactly that problem... :facepalm: 
Sorry for asking this long message. As always many thanks for your fast response and advice!
